source/simulator/solver.cc

```diff
@@ -367,8 +370,135 @@ namespace aspect
+         * @param solver_tolerance The relative solver tolerance for the inner solve
+         * @param inverse_lumped_mass_matrix Lumped mass matrix associated with the velocity block
+         */
+        WeightedBFBT(const TrilinosWrappers::SparseMatrix &mp_matrix,
+                     const PreconditionerMp &mp_preconditioner,
+                     const double solver_tolerance,
+                     const TrilinosWrappers::MPI::Vector &inverse_lumped_mass_matrix,
+                     const TrilinosWrappers::BlockSparseMatrix &system_matrix);
+
+        void vmult(TrilinosWrappers::MPI::Vector &dst,
+                   const TrilinosWrappers::MPI::Vector &src) const;
+
+        unsigned int n_iterations() const;
+
+      private:
+        mutable unsigned int n_iterations_;
+        const TrilinosWrappers::SparseMatrix &mp_matrix;
+        const PreconditionerMp &mp_preconditioner;
+        const double solver_tolerance;
+        const TrilinosWrappers::MPI::Vector  &inverse_lumped_mass_matrix;
+        const TrilinosWrappers::BlockSparseMatrix &system_matrix;
+    };
+
+    template <class PreconditionerMp>
+    WeightedBFBT<PreconditionerMp>::WeightedBFBT(
+      const TrilinosWrappers::SparseMatrix &mp_matrix,
+      const PreconditionerMp &mp_preconditioner,
+      const double solver_tolerance,
+      const TrilinosWrappers::MPI::Vector &inverse_lumped_mass_matrix,
+      const TrilinosWrappers::BlockSparseMatrix &system_matrix)
+      : n_iterations_ (0),
+        mp_matrix (mp_matrix),
+        mp_preconditioner (mp_preconditioner),
+        solver_tolerance (solver_tolerance),
+        inverse_lumped_mass_matrix(inverse_lumped_mass_matrix),
+        system_matrix(system_matrix)
+    {}
+
+
+    template <class PreconditionerMp>
+    void WeightedBFBT<PreconditionerMp>::vmult(TrilinosWrappers::MPI::Vector &dst,
+                                               const TrilinosWrappers::MPI::Vector &src) const
+    {
+      // Trilinos reports a breakdown in case src=dst=0, even though it should return
+      // convergence without iterating. We simply skip solving in this case.
+      SolverControl solver_control(1000, src.l2_norm() * solver_tolerance);
+      PrimitiveVectorMemory<LinearAlgebra::Vector> mem;
+      SolverCG<LinearAlgebra::Vector> solver(solver_control, mem);
+
+      try
+        {
+          TrilinosWrappers::MPI::Vector utmp;
+          utmp.reinit(inverse_lumped_mass_matrix);
+          TrilinosWrappers::MPI::Vector ptmp;
+          ptmp.reinit(src);
+          TrilinosWrappers::MPI::Vector wtmp;
+          wtmp.reinit(inverse_lumped_mass_matrix);
+
+
+          {
+            SolverControl solver_control(5000, 1e-6 * src.l2_norm(), false, true);
+            SolverCG<TrilinosWrappers::MPI::Vector> solver(solver_control);
+
+            solver.solve(mp_matrix,
+                         ptmp,
+                         src,
+                         mp_preconditioner); //(BD^{-1}B^T}^{-1})y2 store in ptmp
```

Comment:
We do not use inline comments. Please move all comments in this function before the line they are describing.